project_git/controller/controller.py

```diff
@@ -0,0 +1,425 @@
+        self._payload = False
+
+    @property
+    def request(self):
+        return self._request
+
+    @property
+    def env(self):
+        return self._request.env
+
+    @property
+    def type(self):
+        return self._type
+
+    @property
+    def token(self):
+        return self._token
+
+    @property
+    def repository(self):
+        return self._repository
+
+    @repository.setter
+    def repository(self, repository):
+        self._repository = repository
+
+    @property
+    def parser(self):
+        return self.env['project.git.payload.parser']
+
+    @property
+    def header(self):
+        if not self._header:
+            self._header = self.parser.parse_header(self)
+        return self._header
+
+    @property
+    def signature(self):
+        return self.header['signature']
+
+    @property
+    def has_signature(self):
+        return 'signature' in self.header
+
+    @property
+    def delivery(self):
+        return self.header['delivery']
+
+    @property
+    def raw_payload(self):
+        return self._raw_payload
+
+    @property
+    def payload(self):
+        if not self._payload:
+            self._payload = self.parser.parse(self)
+        return self._payload
+
+    @property
+    def action_type(self):
+        return self.header['action_type']
+
+    @property
+    def has_action(self):
+        return self.header['action_type']
+
+    @property
+    def action(self):
+        return self.header['action']
+
+    def validate_payload(self, payload):
+        payload_digest = hmac_new(self.repository.secret, payload, True)
+        return digest_compare(payload_digest, self.signature)
+
+    def not_found(self):
+        if isinstance(self._request, http.HttpRequest):
+            return self._request.not_found()
+
+        return json.dumps({"response": "Not found!"})
+
+    def render(self, template_xml_id, values):
+        return self.env.ref(template_xml_id).render(values)
+
+    def render_task_subject(self, values):
+        return self.render(
+            'project_git.task_code_committed_notification_subject', values
+        )
+
+    def render_task_body(self, values):
+        return self.render(
+            'project_git.task_code_committed_notification_body', values
+        )
+
+    def render_orphan_subject(self, values):
+        return self.render(
+            'project_git.orphan_code_committed_notification_subject', values
+        )
+
+    def render_orphan_body(self, values):
+        return self.render(
+            'project_git.orphan_code_committed_notification_body', values
+        )
+
+
+class GitController(http.Controller):
+
+    def process_request(self, context):
+        repository = context.env["project.git.repository"].sudo().search([
+            ("type", "=", context.type),
+            ("odoo_uuid", "=", context.token),
+        ], limit=1)
+
+        if not repository:
+            return context.not_found()
+
+        context.repository = repository
+
```

Comment:
or this

Reply:
I see. I'll get right on it ... thanks :)

Reply:
I've just checked that code with my IDE (PyCharm) which has integrated PEP8 validation and it does not show any warnings. Actually It shows warnings when I add additional empty line here. IMHO code is less readable when lines are that tight together, so I wouldn't change that but if you thing that's necessary I'll do that ....